contrib/babelfishpg_tsql/src/procedures.c

```diff
@@ -365,24 +370,38 @@ sp_describe_first_result_set_internal(PG_FUNCTION_ARGS)
 		attinmeta = TupleDescGetAttInMetadata(tupdesc);
 		funcctx->attinmeta = attinmeta;
 
-		/* If TSQL Query is NULL string then send no rows. */
-		if (batch && batch[0] != '\0')
+		/* First, pass the batch to the ANTLR parser. */
+		if (batch)
 		{
-			/* TODO call ANTLR to parse the query and return NULL for valid non-select queries. */
-			if (strncmp(batch, "select", 6) != 0)
-					ereport(ERROR,
-							 (errcode(ERRCODE_INTERNAL_ERROR),
-							 errmsg("sp_describe_first_result_set supports only select statements %s", query)));
+			result = antlr_parser_cpp(batch);
+			if (!result.success)
+				report_antlr_error(result);
 
-			query = psprintf("CREATE VIEW %s as %s", sp_describe_first_result_set_view_name, batch);
+			/* Skip if NULL query was passed. */
+			if (pltsql_parse_result->body)
+				parsedbatch = ((PLtsql_stmt_execsql *)lsecond(pltsql_parse_result->body))->sqlstmt->query;
+		}
+
+		/* If TSQL Query is NULL string or a non-select query then send no rows. */
+		if (parsedbatch && strncmp(parsedbatch, "select", 6) == 0)
+		{
+			sp_describe_first_result_set_inprogress = true;
+			query = psprintf("CREATE VIEW %s as %s", sp_describe_first_result_set_view_name, parsedbatch);
 	
 			/* Switch Dialect so that SPI_execute creates a TSQL View, obeying TSQL Syntax. */
 			set_config_option("babelfishpg_tsql.sql_dialect", "tsql",
 										(superuser() ? PGC_SUSET : PGC_USERSET),
 											PGC_S_SESSION, GUC_ACTION_SAVE, true, 0, false);
-
 			if ((rc = SPI_execute(query, false, 1)) < 0)
```

Comment:
we should consider a TRY/CATCH as a throw from query execution will leak above settings

Reply:
1. Having try/catch in this function runs into the same as issue as BABEL-3042
2. It is very unlikely that we run into that situation for VALID select queries since we at least validate syntactically while parsing query using ANTLR